### PR TITLE
Add custom icons support

### PR DESCRIPTION
**Pull request by Footmau**
Possibility to read and modify custom icons of groups and entries.
Didn't implemented test yet, I'm thinking about how to do it.

### CODE REVIEW

**Comment by tobischo** [comment]
From which KDBX file version are these icons supported?
What are the restrictions on file format, size etc?

**Comment by Footmau** [comment]
This feature supported in both KDBX 3.1 and 4 versions.
Supported format you can see on the attached screenshot, but I think it is storead as BMP anyway.
No other restrictions I found, probably there are some in implementation, but not documented.

image

**Comment by tobischo** [comment]
As far as I can tell from the code and the example I tested with a random jpg, the data is stored as `.png` and converted accordingly, if another format is provided.
Also it was stored in 128x128px and a resolution of 96x96dpi

It might make sense to also add matching convenience functions for at least some formats to initialize the data accordingly :thinking:

**Comment by Footmau** [comment]
Do you mean convert image to appropriate format?
It can be client responsibility. Maybe just comment with the restrictions will be enought?

**Comment by tobischo** [comment]
The conversion part could be part of whichever client uses this library, that's true.

I suppose a check which ensures the format is matching the requirements would be a good idea.
That could also contain an error message which shows the requirements if they are not met, making it easy to spot what's wrong if the client passes data which does not fullfil them

File: xml.go [hunk 1]
```diff
@@ -71,6 +98,7 @@ type MetaData struct {
 	MasterKeyChangeRec         int64          `xml:"MasterKeyChangeRec"`
 	MasterKeyChangeForce       int64          `xml:"MasterKeyChangeForce"`
 	MemoryProtection           MemProtection  `xml:"MemoryProtection"`
+	CustomIcons                []CustomIcon   `xml:"CustomIcons>Icon"`
```
Comment (tobischo) [inline xml.go]: If no custom icons are defined, this is not part of the xml when it comes from KDBX and therefore should probably be optional

File: xml.go [hunk 2]
```diff
@@ -138,6 +166,7 @@ func WithEntryFormattedTime(formatted bool) EntryOption {
 type Entry struct {
 	UUID            UUID              `xml:"UUID"`
 	IconID          int64             `xml:"IconID"`
+	CustomIconUUID  UUID              `xml:"CustomIconUUID"`
```
Comment (tobischo) [inline xml.go]: This field is not there when no custom icon is set and the XML is marshalled by original KDBX. Therefore this should be optional

File: xml.go [hunk 3]
```diff
@@ -301,6 +337,7 @@ type Group struct {
 	Name                    string        `xml:"Name"`
 	Notes                   string        `xml:"Notes"`
 	IconID                  int64         `xml:"IconID"`
+	CustomIconUUID          UUID          `xml:"CustomIconUUID"`
```
Comment (tobischo) [inline xml.go]: It is the same as for entries:
I no custom icon is defined, the UUID field is not there at all and needs to be set as optional

**Comment by Footmau** [comment]
Thanks for review. I'll fix all the issues a little bit later, no free time for now.
Actually I use this feature as read-only. But I suggest to add a kind of modifiers that will perform necessary checks. Not only for add, but also for remove icon (check that there is no references to this icon).

File: binary.go [hunk 4]
```diff
@@ -66,7 +66,7 @@ func (bs *Binaries) Add(c []byte) *Binary {
 }
 
 // GetContent returns a string which is the plaintext content of a binary
-func (b Binary) GetContent() (string, error) {
+func (b Binary) GetContent() ([]byte, error) {
```
Comment (tobischo) [inline binary.go]: Maybe just to weigh in here on this change:
This would be a breaking change and require a major version update (again..)

Comment (Footmau) [reply binary.go]: Oh, sorry. I didn’t though this change will be added to the merge request.

**Pull request closed without merging**